app/src/main/java/com/javahelp/frontend/domain/user/login/LoginInteractor.java

```diff
@@ -0,0 +1,93 @@
+package com.javahelp.frontend.domain.user.login;
+
+import com.javahelp.model.user.IPasswordHasher;
+import com.javahelp.model.user.User;
+import com.javahelp.model.user.UserPassword;
+
+import org.apache.hc.core5.concurrent.FutureCallback;
+
+/**
+ * Use case interactor for logging in a {@link User}
+ */
+public class LoginInteractor implements ILoginInput {
+
+    ILoginOutput output;
+    ISaltDataAccess saltAccess;
+    ILoginDataAccess loginAccess;
+    IPasswordHasher hasher;
+
+    /**
+     * Creates a new {@link LoginInteractor}
+     *
+     * @param output      {@link ILoginOutput} to output to
+     * @param saltAccess  {@link ISaltDataAccess} to get salt from
+     * @param loginAccess {@link ILoginDataAccess} to login using
+     * @param hasher      {@link IPasswordHasher} to hash password
+     */
+    public LoginInteractor(ILoginOutput output, ISaltDataAccess saltAccess, ILoginDataAccess loginAccess,
+                           IPasswordHasher hasher) {
+        this.output = output;
+        this.saltAccess = saltAccess;
+        this.loginAccess = loginAccess;
+        this.hasher = hasher;
+    }
+
+    @Override
+    public void login(String id, String username, String email, String password, boolean stayLoggedIn) {
+
+        if (username == null && email == null && id == null) {
```

Comment:
Is Id the token id?

Reply:
ID here is the ID of the user to log into.  One of `id`, `username`, and `email` must be non-null. The other two can be null. If all three are null, an error is thrown. This allows users to login using a variety of different pieces of identifying information. If we wanted to change it so that in the future users could log in by their email, we could. Similarly, if we wanted users to be able to save their user IDs and log in faster by just entering their password, we could!